### PR TITLE
Misc bug fixes in Zero optimizer: handling differentiable argument, optimizer_dtype

### PR DESCRIPTION
**Pull request by amithrm**
This a cumulative PR with misc bug fixes and updates to Zero Redundancy Optimizer from all the authors (AWS): Guangtai Huang, Rahul Solanki, Fei Wu, Amith Mamidala

### CODE REVIEW

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 1]
```diff
@@ -91,17 +95,36 @@ def init_zero(self):
...
-          f"Current rank {self.global_rank} is missing from the sharding_groups {self.sharding_groups}"
-      )
+      raise ValueError(f"Current rank {self.global_rank} is missing from the sharding_groups {self.sharding_groups}")
     # Shard parameters for use in optimizer
     sharded_param_groups = self._shard_parameters()
     # Optimizer initialization
+    # Here we pop the differentiable default because the adam family of
+    # optimizers don't have differentiable as an argument. This should
+    # be fixed by this commit https://github.com/pytorch/pytorch/pull/86183
+    # and should be available in torch==2.0. For 1.13, we are patching it here.
```
Comment (jeffhataws) [inline torch_xla/distributed/zero_redundancy_optimizer.py:105]: Hi @amithrm , here it says "This should be fixed by this commit https://github.com/pytorch/pytorch/pull/86183 and should be available in torch==2.0." Can you remove this patch?

Comment (amithrm) [reply torch_xla/distributed/zero_redundancy_optimizer.py:105]: yes

Comment (amithrm) [reply torch_xla/distributed/zero_redundancy_optimizer.py]: @jeffhataws PTAL

**Review by alanwaketan — COMMENTED** [review]
Can we have a test case to cover the change?

File: torch_xla/distributed/zero_redundancy_optimizer.py [hunk 2]
```diff
@@ -296,13 +317,23 @@ def step(self, closure=None, **kwargs):
           shard_data = shard.data
           if param.dtype != self.optimizer_dtype:
             shard_data = shard_data.to(dtype=param.dtype)
-          padded_param = xm.all_gather(
-              shard_data,
-              dim=0,
-              pin_layout=self.pin_layout,
-              groups=self.sharding_groups,
-          )
+          sharded_data.append(shard_data)
```
Comment (jeffhataws) [inline torch_xla/distributed/zero_redundancy_optimizer.py:320]: Is this gathering all the parameters into one bucket?

**Comment by jeffhataws** [comment]
The changes here should already be in https://github.com/pytorch/xla/pull/6025 , as confirmed by Guangtai.

**Pull request closed without merging**